src/couch_mrview/src/couch_mrview_http.erl

```diff
@@ -709,3 +750,169 @@ maybe_set_page_size(#mrargs{page_size = undefined} = Args, MaxPageSize) ->
     Args#mrargs{page_size = MaxPageSize};
 maybe_set_page_size(#mrargs{} = Args, _MaxPageSize) ->
     Args.
+
+
+bookmark_encode(Args0) ->
+    {Args, _} = lists:foldl(fun(Value, {Acc, Idx}) ->
+        case element(Idx, Acc) of
+            Value ->
+                {setelement(Idx, Acc, []), Idx + 1};
+            _Default when Idx == #mrargs.bookmark
+                -> {setelement(Idx, Acc, []), Idx + 1};
+            _Default ->
+                {setelement(Idx, Acc, Value), Idx + 1}
+        end
+    end, {#mrargs{}, 1}, tuple_to_list(Args0)),
+    Term = erlang:delete_element(1, Args),
+    Binary = term_to_binary(Term, [compressed, {minor_version, 2}]),
+    couch_util:encodeBase64Url(Binary).
+
+
+bookmark_decode(Bookmark) ->
+    Binary = couch_util:decodeBase64Url(Bookmark),
+    Term = binary_to_term(Binary, [safe]),
+    {Args, _} = lists:foldl(fun(Value, {Acc, Idx}) ->
+        case {Value, element(Idx, Acc)} of
+            {[], []} ->
+                {Acc, Idx + 1};
+            {[], Default} ->
+                {setelement(Idx, Acc, Default), Idx + 1};
+            {_, _} when Idx == #mrargs.bookmark ->
+                {setelement(Idx, Acc, true), Idx + 1};
+            {Value, _} ->
+                {setelement(Idx, Acc, Value), Idx + 1}
+        end
+    end, {#mrargs{}, 1}, [mrargs | tuple_to_list(Term)]),
+    Args.
+
+
+paginated(Req, EtagTerm, #mrargs{page_size = PageSize} = Args, KeyFun, Fun) ->
+    Etag = couch_httpd:make_etag(EtagTerm),
+    chttpd:etag_respond(Req, Etag, fun() ->
+        hd(paginated(PageSize, [set_limit(Args)], KeyFun, Fun))
+    end).
+
+
+paginated(Req, EtagTerm, PageSize, QueriesArgs, KeyFun, Fun) when is_list(QueriesArgs) ->
+    Etag = couch_httpd:make_etag(EtagTerm),
+    chttpd:etag_respond(Req, Etag, fun() ->
+        Results = paginated(PageSize, QueriesArgs, KeyFun, Fun),
+        #{results => Results}
+    end).
+
+paginated(PageSize, QueriesArgs, KeyFun, Fun) when is_list(QueriesArgs) ->
```

Comment:
If I'm reading this right, in the case of a multi-query HTTP response you're returning a separate bookmark for each individual query, but applying a page limit to across all queries. Assuming I've got that correct, I'm confused how a client would make use of that. Is it expected that we'd POST back all the bookmarks? That seems somewhat odd but I guess the way that our multi-query returns individual responses maybe its the least worst thing?

Also using foldr here seems awkward. If I submit a list of 5 queries that need to be paged, I'd very much expect to page through queries in the order I provided. Doing it in reverse seems less than useful.

Reply:
> you're returning a separate bookmark for each individual query, but applying a page limit to across all queries.

Yes I am. Otherwise the client could bypass the limit by using huge number of queries. Currently (in this PR) we limit number of queries in POST body to 2000. Even with this setting if we would apply individual limits the user would get 2000^2=4000000 rows in a single response. An alternative approach is to set individual limits to be `2000/n`. Where `n` is number of queries in the POST body. The downside of this approach is: it makes reusing of bookmarks in separate _all_docs requests less efficient. Currently client can issue POST, get a bunch of bookmarks (each bookmark has `page_size=2000`) and call GET with each bookmark.

> Is it expected that we'd POST back all the bookmarks?

It is possibility in case when client use a certain combination of parameters. For example when the client set `page_size` equal to number of queries in queries. However these bookmarks can be used in individual calls using GET.

Reply:
I'm struggling to envision how a client would use this API in a more efficient or ergonomic way than just issue each query individually. My personal expectation ignoring the current implementation would be that I a pagination response would return result sets in the order I specified them up to one page at a time. And getting the next page would just be a matter of passing a single bookmark.

One oddity of that approach is that I'm not sure what the response format would look like after you paged through the initial result sets. Maybe it'd return an empty response or something that indicates "already consumed" or some such?

Alternatively as you suggest maybe spreading the page limit across all queries with the `2000/n` approach? But that still doesn't cover what happens if any of the queries result sets are exhausted before all of the others.

Reply:
You can find some additional background info in the RFC https://github.com/apache/couchdb-documentation/pull/530/files#diff-062d330b3595274c2461c38a43f29a11R135